src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -684,7 +684,16 @@ bool primitive_inst::use_async_compilation() {
         return false;
     }
 
-    return (_node->is_type<convolution>() || _node->is_type<fully_connected>() || _node->is_type<gemm>() ||
+    bool compile_fc_impls = _node->is_type<fully_connected>();
+    if (compile_fc_impls) {
+        const auto& fc_node = _node->as<fully_connected>();
+        if (fc_node.get_primitive()->compressed_weights) {
+            auto weights_dt = fc_node.weights().get_output_layout().data_type;
+            compile_fc_impls = !one_of(weights_dt, {data_types::i4, data_types::u4});
+        }
+    }
```

Comment:
1. Previously when I tested with i8, async compiled fc was 20% better than shape agnostic kernel. Is it different for fc int4?
2. Also kernel reusability of fc is very high, we only need 3~4 fcs for the generative models' run. Still not valuable?

Reply:
Thanks for the comments!
1. Yes, as I can see, the situation for int4 looks different and performance is on par or even slightly better for the most of the cases
2. The main reason of disabling async compilation for int4 is performance optimization. Current approach works fine for 2nd tokens while we have num_beams=1 (in that case we use optimized shape agnostic kernel and may have async-compiled static kernel for batch=1).
But the issue is happening in case of num_beams>1. For instance, for num_beams=3 we use default shape-agnostic kernel version which uses TILE_B=8 (and then we calculate extra 5 batches) - this can be solved with my changes in current PR. But we also have fake_alignment by 8 for all cases where batch > 1 - and this alignment applies to static kernel's shape. So, async-compiled kernel with TILE_B=8 will be used for all batches between 2 and 8 - and again we have to calculate extra 5 batches in static kernel version
There were two options: a) disable async compilation at all; b) decrease granulation of fake alignment and static kernels shapes for batches < 8
I preferred a) as it allows to implement simpler logic w/o extra complication and additional fake alignment and kernels compilation rules changes

Reply:
@isanghao, as I can see, there is no negative performance impact
Please find details in related Jira ticket - 129757